connectivity/service/src/main/java/org/eclipse/ditto/connectivity/service/mapping/CloudEventsMapper.java

```diff
@@ -0,0 +1,166 @@
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.ditto.connectivity.service.mapping;
+
+import static java.util.Collections.singletonList;
+import static org.eclipse.ditto.json.JsonFactory.newObject;
+
+import java.io.UnsupportedEncodingException;
+import java.util.*;
+
+import com.eclipsesource.json.Json;
+import org.eclipse.ditto.base.model.common.DittoConstants;
+import org.eclipse.ditto.base.model.exceptions.DittoJsonException;
+import org.eclipse.ditto.base.model.exceptions.DittoRuntimeException;
+import org.eclipse.ditto.base.model.headers.DittoHeaders;
+import org.eclipse.ditto.connectivity.api.ExternalMessage;
+import org.eclipse.ditto.connectivity.api.ExternalMessageFactory;
+import org.eclipse.ditto.connectivity.model.ConnectivityModelFactory;
+import org.eclipse.ditto.connectivity.model.MappingContext;
+import org.eclipse.ditto.connectivity.model.MessageMappingFailedException;
+import org.eclipse.ditto.internal.utils.akka.logging.DittoLogger;
+import org.eclipse.ditto.internal.utils.akka.logging.DittoLoggerFactory;
+import org.eclipse.ditto.json.*;
+import org.eclipse.ditto.messages.model.MessageHeaders;
+import org.eclipse.ditto.protocol.Adaptable;
+import org.eclipse.ditto.protocol.JsonifiableAdaptable;
+import org.eclipse.ditto.protocol.ProtocolFactory;
+
+/**
+ * A message mapper implementation for the Mapping incoming CloudEvents to Ditto Protocol.
+ */
+@PayloadMapper(alias = {"CloudEvents",
+    // legacy full qualified name
+    "org.eclipse.ditto.connectivity.service.mapping.CloudEventsMapper"})
+public final class CloudEventsMapper extends AbstractMessageMapper {
+
+
+  static final String CE_ID = "ce-id";
+  static final String CE_TYPE = "ce-type";
+  static final String CE_SOURCE = "ce-source";
+  static final String CE_SPECVERSION = "ce-specversion";
+  static final String CONTENT_TYPE = "application/cloudevents+json";
+  static final String SPECVERSION = "specversion";
+  static final String ID = "id";
+  static final String SOURCE = "source";
+  static final String TYPE = "type";
+  static final String OUTBOUNDTYPE = "org.eclipse.ditto.outbound";
+  static final String OUTBOUNDSPECVERSION = "1.0";
+  static final String OUTBOUNDSOURCE = "org.eclipse.ditto";
+
+
+  @Override
+  public List<Adaptable> map(final ExternalMessage message) {
+    MessageMappingFailedException mappingFailedException = MessageMappingFailedException.newBuilder(
+            message.findContentType().orElse("")).description("This is not a CloudEvent")
+        .dittoHeaders(DittoHeaders.of(message.getHeaders())).build();
+    if (message.findContentType().orElse("").equals(CONTENT_TYPE)) {
```

Comment:
Must a "binary CloudEvent" really contain the CloudEvent contentType?
As I understood the "binary CloudEvent" the CloudEvent attributes are read from headers whereas the payload does not have anything to do with a CloudEvent.
Or am I wrong?

Reply:
Yes, a cloud event always has a content-type:
* https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/bindings/kafka-protocol-binding.md#321-content-type
* https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/bindings/kafka-protocol-binding.md#331-kafka-content-type

In the case of a binary cloud event, the content type is mapped to the "data content type", in the structural cloud event is must have a value of `application/cloudevents+json; charset=UTF-8` (or `application/cloudevents+json`).